src/rpc/core_rpc_server.cpp

```diff
@@ -1111,9 +1232,74 @@ namespace cryptonote
     return true;
   }
   //------------------------------------------------------------------------------------------------------------------------------
+  template <typename COMMAND_TYPE>
+  bool core_rpc_server::use_bootstrap_daemon_if_necessary(const std::string &type, const std::string &command_name, const typename COMMAND_TYPE::request& req, typename COMMAND_TYPE::response& res, bool &r)
+  {
+    res.untrusted = false;
+    if (m_bootstrap_daemon_address.empty())
+      return false;
+
+    boost::lock_guard<boost::mutex> lock(m_bootstrap_daemon_mutex);
+    if (!m_should_use_bootstrap_daemon)
+    {
+      MINFO("The local daemon is fully synced. Not switching back to the bootstrap daemon");
+      return false;
+    }
+
+    auto current_time = std::chrono::system_clock::now();
+    if (current_time - m_bootstrap_height_check_time > std::chrono::seconds(30))  // update every 30s
+    {
+      m_bootstrap_height_check_time = current_time;
+
+      uint64_t top_height;
+      crypto::hash top_hash;
+      m_core.get_blockchain_top(top_height, top_hash);
+      ++top_height; // turn top block height into blockchain height
+      uint64_t target_height = m_core.get_target_blockchain_height();
+
+      m_should_use_bootstrap_daemon = target_height ? top_height < target_height : true;
+      MINFO((m_should_use_bootstrap_daemon ? "Using" : "Not using") << " the bootstrap daemon (top height: " << top_height << ", target height: " << target_height << ")");
+    }
+    if (!m_should_use_bootstrap_daemon)
+      return false;
+
+    if (type == "JON")
```

Comment:
`JSON`

Reply:
Don't use magic strings, use an enum or something similar.

Reply:
It's simpler, and this function isn't meant to be called from outside.

Reply:
ok